lib/elixir/lib/access.ex

```diff
@@ -185,37 +199,40 @@ defmodule Access do
   @doc """
   Invoked in order to access the value under `key` and update it at the same time.
 
-  The implementation of this callback should invoke the passed function with the
-  value under key `key` in the passed structure, or `nil` if the key is not
-  present. This function should return either `{value_to_return, new_value}` or
-  `:pop`.
+  The implementation of this callback should invoke `fun` with the value under
+  key `key` in the passed `structure`, or with `nil` if `key` is not present in it.
+  This function must return either `{get_value, update_value}` or `:pop`.
 
-  If the passed function returns `{value_to_return, new_value}`, the return
-  value of this callback should be `{value_to_return, new_term}` where
-  `new_term` is `term` after updating the value of `key` with `new_value`.
+  If the passed function returns `{get_value, update_value}`,
+  the return value of this callback should be `{get_value, new_structure}`, where:
+  - `get_value` is the retrieved value (which can be operated on before being returned)
+  - `update_value` is the new value to be stored under `key`
+  - `new_structure` is `structure` after updating the value of `key` with `update_value`.
 
   If the passed function returns `:pop`, the return value of this callback
-  should be `{value, new_term}` where `value` is the value under `key` or `nil`
-  if not present, and `new_term` is `term` without the key `key`.
+  must be `{value, new_structure}` where `value` is the value under `key`
+  (or `nil` if not present) and `new_structure` is `structure` without the key `key`.
 
   See the implementations of `Map.get_and_update/3` or `Keyword.get_and_update/3`
   for more examples.
   """
-  @callback get_and_update(term :: t, key, (value -> {value, value} | :pop)) :: {value, t}
+  @callback get_and_update(structure, key, fun :: (value | nil -> {get_value, update_value} | :pop)) ::
```

Comment:
Those specs are more precise but they are also much harder to read. There are too many variables. And given we use typespecs mostly for readability purposes, I don't think this is a positive change.

I would propose:

    @callback get_and_update(data, key, (value -> {get_value, update_value} | :pop)) :: {get_value, data} when
              data: container, get_value: value, update_value: value

Don't worry about the nil cases, they were already included when we listed the type as `any`.